### PR TITLE
fix(profiling): `asyncio` task fixes for stack v2

### PR DESCRIPTION
**Pull request by taegyunkim**
This PR fixes 2 issues.

1. Associate `asyncio` tasks with spans

Task rendering begins with calling `render_task_begin`, and it is then followed by call to `FrameStack::render()`. However, we only add span id association via call to `render_thread_begin()` so tasks are missing span id labels. To properly link spans to tasks, we add the label when we create a new sample for task. And while debugging this issue, discovered the following. 

2. Propagate `asyncio` task names properly

Echion adds a dummy frame having task name, code [here](https://github.com/taegyunkim/echion/blob/9d5bcc5867d7aefff73c837adcba4ef46eecebc6/echion/threads.h#L282). When it tries to render tasks, it [peeks at the top of the frame](https://github.com/taegyunkim/echion/blob/9d5bcc5867d7aefff73c837adcba4ef46eecebc6/echion/threads.h#L351) stack's name and uses that. However, there could be frames preceding the frame stack added by [this line](https://github.com/taegyunkim/echion/blob/9d5bcc5867d7aefff73c837adcba4ef46eecebc6/echion/threads.h#L276). Leading to using incorrect task names.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (sanchda) [inline ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:135]: How do you feel about skipping the checks for invalid names?  We already propagate invalid symbol and filenames.

Actually, the UX for that might not be great, since it'll erroneously associate a bunch of execution into potentially invalid tasks (like we might have hundreds of tasks all associated to "<invalid>").

Instead:  what if we have a static sequence number that (atomically) increments every time we see an invalid task, and name the task something like "invalid-XXX"?  That'd eliminate the need to check, it'd make sure tasks are labeled when execution is associated to a task (this is important!), and it'd isolate execution away from a common label.

File: releasenotes/notes/profiling-stack-v2-task-names-1d566a1793280aa0.yaml
Comment (sanchda) [inline releasenotes/notes/profiling-stack-v2-task-names-1d566a1793280aa0.yaml:6]: Copy/paste issue?

File: tests/profiling_v2/collector/test_stack_asyncio.py
Comment (sanchda) [inline tests/profiling_v2/collector/test_stack_asyncio.py:62]: If we dump out before on python 3.7, we can probably lose this branch as well

**Review by sanchda — APPROVED** [review]
Very important 👍

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (nsrip-dd) [inline ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:69]: It's really not obvious to me just from the PR description or any comments in the code why this wasn't right before or why it's right now. As far as I can piece it together, the name here is derived when Echion enumerates tasks during a sample call. Is it not computing the name correctly? Are we adding the same name multiple times? Could you please include some description of the bug and why this fix works? And do we need to fix anything in your Echion fork, given that it's passing a no-longer-useful parameter here?

Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:69]: It is not computing the name correctly. Updated the PR description. I don't think we need to fix anything in the Echion fork yet.

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:97]: I am not sure whether I understood your question correctly. From my understanding, there's 1-to-many relationship between threads and tasks, and we have 1-to-1 relationship between threads and spans. So, I'm just connecting the dots to associate span ids to tasks. 

We have `asyncio` machinery defined [here](https://github.com/DataDog/dd-trace-py/blob/67eb77dacca1d8679ca7fa088e3c873cba840208/ddtrace/profiling/_asyncio.py#L33) for both v1 and v2, btw.

File: releasenotes/notes/profiling-stack-v2-task-names-1d566a1793280aa0.yaml
Comment (taegyunkim) [reply releasenotes/notes/profiling-stack-v2-task-names-1d566a1793280aa0.yaml:6]: Oh this PR fixes these two issues.

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (P403n1x87) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:135]: There's always the option of "fixing" echion to report better metadata. For instance, we could make the frame object carry a `kind`, that we could set to a value representing "Task", instead of working around the shortcoming?

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:138]: We could. Do we want or not want to push the task name as a frame?

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp]: Removed `file`. I think I added the checks as I've been getting `<invalid>` for task names in the tests. I don't mind showing more of `<invalid>` ask we're already showing this to the users.

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp]: Let me try removing those checks. 

@P403n1x87's suggestion would be the cleanest way here. Propagate a flag from echion and check that instead of relying on `line == 0`. I intentionally avoided changing echion code as I didn't want it to diverge more.

Comment (sanchda) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp]: Rather than propagating new information in the underlying struct, it might be cleaner to structure the interface to push a different category of information.

We'd have a `render_task_name` interface in the echion fork's Render system.  The Echion implementation of Render would just push it as a frame, and ddtrace would push it into ddup as a task name.

File: releasenotes/notes/profiling-stack-v2-task-names-1d566a1793280aa0.yaml
Comment (sanchda) [reply releasenotes/notes/profiling-stack-v2-task-names-1d566a1793280aa0.yaml:6]: Oh, sorry.  I saw the single `- |` and forgot how these yamls are parsed.  Carry on, apologies for the noise

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:138]: While I _do_ really like the idea of having the tasks all visualized in the flamegraph, it does conflict with our UX at the moment.

If we want to retain the desired output, I think it'd be really interesting to see whether we can add the pseudo-frames in the backend.  That way we'd be able to split things at the root level in a bunch of different ways.

So my uninformed and weakly held opinion is that we populate the task name, but do not submit this as a frame.

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp]: Yeah--I don't think there's a wrong way or a right way.

The only quantitative thing here is that omitting the task name leads to a UX where the underlying execution appears as though it did not occur within a task at all.  It's actually been incredibly useful (for, uh... profiler escalations 😢) to know when things are on a task or not--even if the task name is garbage.

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:138]: OK, yeah we also have this side panel selector for task names. 

Screenshot 2024-12-04 at 11 58 45 AM

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp]: That's a reasonable concern but it would be better to be implemented in echion. Instead of simply using `<invalid>`, it could use context specific `<invalid>`, i.e. <`invalid-task-name>`

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp]: Yeah, we should update echion. I'm just delaying such changes to happen after we merge our fork to the upstream echion.

File: ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp
Comment (nsrip-dd) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:97]: Right, I was wondering about the span-to-task/coroutine relationship. I think I see how it works though. We get the active span for a thread through the tracer's context provider: https://github.com/DataDog/dd-trace-py/blob/47c7b5287da25643e46652e6d222a40a52f2382a/ddtrace/profiling/collector/stack.pyx#L527. And I guess it's up to the tracer to update the context provider when a task/coroutine related to a span becomes active on a thread. So it's not directly related to this PR, just curious how it works :)

Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp:97]: Oh right, you found the right code location. The tracer simply let us know the span when it's activated, and we use that information to keep track of span <-> thread relation

**Review by sanchda — APPROVED** [review]

**Pull request merged**
